// @vite-plugin-vercel/vike/dist/templates/edge-helpers.js
import {
  getOriginalUrl
} from "../chunk-UINMFBLZ.js";

// templates/edge-helpers.ts
import { parse } from "qs";
function getDefaultPageContextInit(request) {
  const queryString = new URL(request.url).search.slice(1);
  const query = parse(queryString);
  const url = getOriginalUrl(request.headers.get("x-now-route-matches"), query.__original_path, request.url);
  const cookieHeader = request.headers.get("Cookie") || "";
  const cookies = cookieHeader.split("; ").reduce(
    (acc, cookie) => {
      const [name, value] = cookie.split("=");
      acc[name] = value;
      return acc;
    },
    {}
  );
  return {
    urlOriginal: url,
    body: request.body,
    cookies
  };
}
function getDefaultEmptyResponseHandler() {
  return new Response(null, {
    status: 200,
    headers: {
      "content-type": "text/html; charset=UTF-8"
    }
  });
}
function getDefaultResponseHandler(httpResponse) {
  const { statusCode, body, headers } = httpResponse;
  return new Response(body, {
    status: statusCode,
    headers
  });
}
export {
  getDefaultEmptyResponseHandler,
  getDefaultPageContextInit,
  getDefaultResponseHandler
};
